### PR TITLE
[tutorial] Add back eager mode static quant tutorial as rst file

### PR DESCRIPTION
**Pull request by jerryzh168**

### CODE REVIEW

File: advanced_source/static_quantization_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,675 @@
+(beta) Static Quantization with Eager Mode in PyTorch 
+========================================================= 
+**Author**: `Raghuraman Krishnamoorthi <https://github.com/raghuramank100>`_
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:3]: Please add your name to the list of authors

File: advanced_source/static_quantization_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+
+3. Define dataset and data loaders  
+----------------------------------  
+ 
+As our last major setup step, we define our dataloaders for our training and testing set. 
+ 
+ImageNet Data 
+^^^^^^^^^^^^^ 
+
+The specific dataset we've created for this tutorial contains just 1000 images from the ImageNet data, one from
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:311]: Let us state something like " Please download imagenet dataset by following the instructions at: TBD. The rest of the tutorial assumes that you have imagenet data available at the location ~/data/imagenet"

File: advanced_source/static_quantization_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+
+    print("Size of baseline model") 
+    print_size_of_model(float_model)  
+
+    top1, top5 = evaluate(float_model, criterion, data_loader_test, neval_batches=num_eval_batches) 
+    print('Evaluation accuracy on %d images, %2.2f'%(num_eval_batches * eval_batch_size, top1.avg)) 
+    torch.jit.save(torch.jit.script(float_model), saved_model_dir + scripted_float_model_file)
+
+  
+We see 78% accuracy on 300 images, a solid baseline for ImageNet,
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:436]: Let us change these numbers to full accuracy on imagenet eval data. We can also add that measuring the full accuracy takes about 5 minutes on a GPU.

File: advanced_source/static_quantization_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+    # Note fusion of Conv+BN+Relu and Conv+Relu 
+    print('\n Inverted Residual Block: After fusion\n\n',float_model.features[1].conv)  
+
+  
+Finally to get a "baseline" accuracy, let's see the accuracy of our un-quantized model  
+with fused modules  
+
+.. code:: python
+
+    num_eval_batches = 10
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:426]: Let us go with num_eval_batches = 1000 and eval_batch_size of 50, to correspond to full imagenet validation dataset size.

File: advanced_source/static_quantization_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+(specifically, this is done by inserting `observer` modules at different points that record this  
+data). These distributions are then used to determine how the specifically the different activations  
+should be quantized at inference time (a simple technique would be to simply divide the entire range  
+of activations into 256 levels, but we support more sophisticated methods as well). Importantly,  
+this additional step allows us to pass quantized values between operations instead of converting these  
+values to floats - and then back to ints - between every operation, resulting in a significant speed-up.  
+
+.. code:: python
+
+    num_calibration_batches = 10
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:456]: Let us change to a realistic number (32 batches)

File: advanced_source/static_quantization_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+
+    torch.quantization.prepare(per_channel_quantized_model, inplace=True) 
+    evaluate(per_channel_quantized_model,criterion, data_loader, num_calibration_batches) 
+    torch.quantization.convert(per_channel_quantized_model, inplace=True) 
+    top1, top5 = evaluate(per_channel_quantized_model, criterion, data_loader_test, neval_batches=num_eval_batches) 
+    print('Evaluation accuracy on %d images, %2.2f'%(num_eval_batches * eval_batch_size, top1.avg)) 
+    torch.jit.save(torch.jit.script(per_channel_quantized_model), saved_model_dir + scripted_quantized_model_file)
+
+  
+Changing just this quantization configuration method resulted in an increase
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:517]: Let us quote the actual accuracies on full imagenet here.

File: advanced_source/static_quantization_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+Training a quantized model with high accuracy requires accurate modeling of numerics at 
+inference. For quantization aware training, therefore, we modify the training loop by:  
+
+- Switch batch norm to use running mean and variance towards the end of training to better  
+  match inference numerics. 
+- We also freeze the quantizer parameters (scale and zero-point) and fine tune the weights. 
+
+.. code:: python
+
+    num_train_batches = 20
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:601]: We can state here that QAT takes time and one needs to train over a few epochs.

File: advanced_source/static_quantization_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+            # Freeze batch norm mean and variance estimates 
+            qat_model.apply(torch.nn.intrinsic.qat.freeze_bn_stats) 
+
+        # Check the accuracy after each epoch 
+        quantized_model = torch.quantization.convert(qat_model.eval(), inplace=False) 
+        quantized_model.eval()  
+        top1, top5 = evaluate(quantized_model,criterion, data_loader_test, neval_batches=num_eval_batches)  
+        print('Epoch %d :Evaluation accuracy on %d images, %2.2f'%(nepoch, num_eval_batches * eval_batch_size, top1.avg)) 
+ 
+Here, we just perform quantization-aware training for a small number of epochs. Nevertheless,
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:619]: Again, we can remove this part and refer to our actual accuracy numbers.

File: advanced_source/static_quantization_tutorial.rst (hunk 1 above)
Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:3]: sure, I added myself to "edited by"

File: advanced_source/static_quantization_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,671 @@
...
+            # Freeze batch norm mean and variance estimates 
+            qat_model.apply(torch.nn.intrinsic.qat.freeze_bn_stats) 
+
+        # Check the accuracy after each epoch 
+        quantized_model = torch.quantization.convert(qat_model.eval(), inplace=False) 
+        quantized_model.eval()  
+        top1, top5 = evaluate(quantized_model,criterion, data_loader_test, neval_batches=num_eval_batches)  
+        print('Epoch %d :Evaluation accuracy on %d images, %2.2f'%(nepoch, num_eval_batches * eval_batch_size, top1.avg)) 
+ 
+Quantization-aware training yields an accuracy of over 71% on the entire imagenet dataset, which is close to the floating point accuracy of 71.9%.
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:617]: We can also mention that we can get even better accuracies when we do QAT for mobile backends and point to https://github.com/pytorch/vision/blob/master/torchvision/models/quantization/mobilenetv2.py

File: advanced_source/static_quantization_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,671 @@
...
+    print('Post Training Quantization: Convert done') 
+    print('\n Inverted Residual Block: After fusion and quantization, note fused modules: \n\n',myModel.features[1].conv) 
+
+    print("Size of model after quantization") 
+    print_size_of_model(myModel)  
+
+    top1, top5 = evaluate(myModel, criterion, data_loader_test, neval_batches=num_eval_batches) 
+    print('Evaluation accuracy on %d images, %2.2f'%(num_eval_batches * eval_batch_size, top1.avg))
+  
+For this quantized model, we see a significantly lower accuracy of just ~62% on these same 300
```
Comment (jerryzh168) [inline advanced_source/static_quantization_tutorial.rst:486]: hi @raghuramank100 is this 67.3% on real dataset?

File: advanced_source/static_quantization_tutorial.rst (hunk 6 above)
Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:514]: is this 67.3% as given in your spreadsheet?

File: advanced_source/static_quantization_tutorial.rst (hunk 9 above)
Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:617]: is this 71.5%?

File: advanced_source/static_quantization_tutorial.rst (hunk 10 above)
Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:486]: Yes, we can remove the lower accuracy of ~62% and say " For this quantized model, we see an accuracy of 67.3% on the eval dataset

File: advanced_source/static_quantization_tutorial.rst (hunk 6 above)
Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:514]: Yes, with the default fbgemm qconfig, we see a big improvement due to per-channel for weights and histogram observers

File: advanced_source/static_quantization_tutorial.rst (hunk 9 above)
Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:617]: Yes, that is correct.

File: advanced_source/static_quantization_tutorial.rst (hunk 6 above)
Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:514]: What is the baseline for this one?

Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:514]: 71.9%?

File: advanced_source/static_quantization_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,665 @@
...
+
+    print("Size of baseline model") 
+    print_size_of_model(float_model)  
+
+    top1, top5 = evaluate(float_model, criterion, data_loader_test, neval_batches=num_eval_batches) 
+    print('Evaluation accuracy on %d images, %2.2f'%(num_eval_batches * eval_batch_size, top1.avg)) 
+    torch.jit.save(torch.jit.script(float_model), saved_model_dir + scripted_float_model_file)
+
+  
+We see 78% accuracy on 300 images, a solid baseline for ImageNet, especially considering our model is just 14.0 MB. Measurnig the full accuracy takes about 5 minutes on a GPU.
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:430]: On the entire model, we get an accuracy of 71.9%

File: advanced_source/static_quantization_tutorial.rst (hunk 6 above)
Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:508]: Yes, I also added a comment earlier, we need to specify the accuracy of the fp32 model.

File: advanced_source/static_quantization_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+
+3. Define dataset and data loaders  
+----------------------------------  
+ 
+As our last major setup step, we define our dataloaders for our training and testing set. 
+ 
+ImageNet Data 
+^^^^^^^^^^^^^ 
+
+Please download imagenet dataset by following the instructions at: TBD. The rest of the tutorial assumes that you have imagenet_1k.zip which contains the sample imagenet dataset.
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:307]: This part "Please download imagenet dataset by following the instructions at: TBD. The rest of the tutorial assumes that you have imagenet_1k.zip which contains the sample imagenet dataset. Unzip imagenet_1k.zip to `data` folder." can be deleted

File: advanced_source/static_quantization_tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+the data using ``torchvision`` following  
+`here <https://pytorch.org/docs/stable/torchvision/datasets.html#imagenet>`_. For example,  
+to download the training set and apply some standard transformations to it, you could use:  
+
+.. code:: python  
+
+    import torchvision  
+    import torchvision.transforms as transforms 
+
+    imagenet_dataset = torchvision.datasets.ImageNet(
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:325]: This part looks incomplete. We are defining an imagenet_dataset first, but not using it later. Also, the transforms here do not match the transforms defined below.

File: advanced_source/static_quantization_tutorial.rst [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+
+        return data_loader, data_loader_test  
+
+
+Next, we'll load in the pre-trained MobileNetV2 model. We provide the URL to download the data from in ``torchvision``  
+`here <https://github.com/pytorch/vision/blob/master/torchvision/models/mobilenet.py#L9>`_. 
+
+.. code:: python
+
+    data_path = 'data/imagenet_1k'
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:388]: Lets replace this from:
data_path = 'data/imagenet_1k'
to
data_path = '~/.data/imagenet'

File: advanced_source/static_quantization_tutorial.rst (hunk 13 above)
Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:325]: this is just used to download training data I think?

Also I didn't see how the transformation is different? can you elaborate?

Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:325]: imagenet_dataset is not being used later, is that the intent?.

Comment (jerryzh168) [reply advanced_source/static_quantization_tutorial.rst:325]: I think this is used to download data

File: advanced_source/static_quantization_tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+.. code:: python
+
+    def prepare_data_loaders(data_path):  
+
+        traindir = os.path.join(data_path, 'train') 
+        valdir = os.path.join(data_path, 'val') 
+        normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],  
+                                         std=[0.229, 0.224, 0.225]) 
+
+        dataset = torchvision.datasets.ImageFolder(
```
Comment (jerryzh168) [inline advanced_source/static_quantization_tutorial.rst:351]: @raghuramank100 we can only read from folder after the data is downloaded

Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:347]: I think we can do this:
'First ask user to download the imagenet archive directly from imagenet, by following the instructions at "http://www.image-net.org/download". Once that is done, we define dataloaders as follows:
```
     def prepare_data_loaders(data_path):  
       normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],  
                                         std=[0.229, 0.224, 0.225])
 dataset = ImageNet(
        data_path, split="train",
      transforms.Compose([  
                transforms.RandomResizedCrop(224),  
                transforms.RandomHorizontalFlip(),  
                transforms.ToTensor(),  
                normalize,  
            ]))  
    dataset_test = ImageNet(
        root_path, split="val", 
            transforms.Compose([  
                transforms.Resize(256), 
                transforms.CenterCrop(224), 
                transforms.ToTensor(),  
                normalize,  
            ])) 
```

Comment (raghuramank100) [reply advanced_source/static_quantization_tutorial.rst:327]: I think we need to use torchvision.datasets.ImageNet because this does some additional work.

File: advanced_source/static_quantization_tutorial.rst [hunk 16]
```diff
@@ -0,0 +1,641 @@
...
+With the data downloaded, we show functions below that define dataloaders we'll use to read 
+in this data. These functions mostly come from  
+`here <https://github.com/pytorch/vision/blob/master/references/detection/train.py>`_.
+
+
+.. code:: python
+
+    def prepare_data_loaders(data_path):  
+
+        traindir = os.path.join(data_path, 'train')
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:322]: We do not need traindir and valdir anymore

File: advanced_source/static_quantization_tutorial.rst [hunk 17]
```diff
@@ -0,0 +1,641 @@
...
+
+3. Define dataset and data loaders  
+----------------------------------  
+ 
+As our last major setup step, we define our dataloaders for our training and testing set. 
+ 
+ImageNet Data 
+^^^^^^^^^^^^^ 
+
+Please download imagenet dataset by following the instructions at: TBD. The rest of the tutorial assumes that you have imagenet_1k.zip which contains the sample imagenet dataset.
```
Comment (raghuramank100) [inline advanced_source/static_quantization_tutorial.rst:307]: Replace" Please download imagenet dataset by following the instructions at: TBD. The rest of the tutorial assumes that you have imagenet_1k.zip which contains the sample imagenet dataset.Unzip imagenet_1k.zip to `data` folder. To run the code in this tutorial using the entire ImageNet dataset, please follow the instructions at `here <http://www.image-net.org/download>`_." 

with

" To run the code in this tutorial using the entire ImageNet dataset, first download imagenet by following the instructions at `here <http://www.image-net.org/download>`_. Unzip the downloaded file into the 'data_path' folder."

**Review by raghuramank100 — APPROVED** [review]

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 18]
```diff
@@ -54,14 +54,9 @@ Advantages of FX Graph Mode Quantization are:
...
-  
-Download dataset: 
-  
-.. code:: 
-  
-    wget https://s3.amazonaws.com/pytorch-tutorial-assets/imagenet_1k.zip 
-  
-and unzip to `data` folder.   
+
+Get a sample imagenet dataset imagenet_1k.zip unzip to `data` folder. (The download url is removed due to legal issues. We'll write down instructions on how to get imagenet_1k.zip data later.)
```
Comment (jspisak) [inline prototype_source/fx_graph_mode_ptq_static.rst:58]: How about:

To avoid requiring the download of the ImageNet dataset, we provide a pointer to download a pretrained model as a starting point.

**Review by jspisak — COMMENTED** [review]
Provided some proposed alternative language. Let me know if this sounds reasonable..

File: prototype_source/fx_graph_mode_ptq_static.rst (hunk 18 above)
Comment (jerryzh168) [reply prototype_source/fx_graph_mode_ptq_static.rst]: thanks, I copied the wording from the eager mode tutorial, please let me know if works

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 19]
```diff
@@ -54,14 +54,9 @@ Advantages of FX Graph Mode Quantization are:
...
-Download dataset: 
-  
-.. code:: 
-  
-    wget https://s3.amazonaws.com/pytorch-tutorial-assets/imagenet_1k.zip 
-  
-and unzip to `data` folder.   
+
+To run the code in this tutorial using the entire ImageNet dataset, first download imagenet by following the instructions at here `ImageNet Data <http://www.image-net.org/download>`_. Unzip the downloaded file into the 'data_path' folder.
+
```
Comment (jspisak) [inline prototype_source/fx_graph_mode_ptq_static.rst:59]: Are we asking the reader to use a pretraining model or download and train on full imagenet? It looks like both at this point.

Comment (jerryzh168) [reply prototype_source/fx_graph_mode_ptq_static.rst:59]: reader will use a pretrained model and do post training quantization, we are downloading the full imagenet just for evaluating the accuracy

Comment (jspisak) [reply prototype_source/fx_graph_mode_ptq_static.rst:59]: got it - thanks Jerry. LGTM..

**Review by jspisak — APPROVED** [review]
